### PR TITLE
Avoid validation on copy of DissimilarityMatrix and DistanceMatrix objects  and validation optimization #1745

### PR DESCRIPTION
**Pull request by sfiligoi**
Speed up copy of DissimilarityMatrix and DistanceMatrix objects, closes issue #1745.

ENH: Avoid validation on copy of DissimilarityMatrix and DistanceMatrix objects.
ENH: Use an  optimized version of symmetry check in DistanceMatrix.

On a MacBook Pro, the creation of a 25k*25k DistanceMatrix goes from about 10s to less than 2s.

Please complete the following checklist:

* [x] I have read the guidelines in [CONTRIBUTING.md](https://github.com/biocore/scikit-bio/blob/master/CONTRIBUTING.md).

* [x] I have documented all public-facing changes in [CHANGELOG.md](https://github.com/biocore/scikit-bio/blob/master/CHANGELOG.md).

* [ ] **This pull request includes code, documentation, or other content derived from external source(s).** If this is the case, ensure the external source's license is compatible with scikit-bio's [license](https://github.com/biocore/scikit-bio/blob/master/COPYING.txt). Include the license in the `licenses` directory and add a comment in the code giving proper attribution. Ensure any other requirements set forth by the license and/or author are satisfied. **It is your responsibility to disclose code, documentation, or other content derived from external source(s).** If you have questions about whether something can be included in the project or how to give proper attribution, include those questions in your pull request and a reviewer will assist you.

* [x] **This pull request does not include code, documentation, or other content derived from external source(s).**

**Note:** [REVIEWING.md](https://github.com/biocore/scikit-bio/blob/master/REVIEWING.md) may also be helpful to see some of the things code reviewers will be verifying when reviewing your pull request.

### CODE REVIEW

File: skbio/stats/distance/_base.py [hunk 1]
```diff
@@ -853,6 +872,38 @@ def __getitem__(self, index):
         else:
             return self.data.__getitem__(index)
 
+    def _validate_ids(self, data, ids):
```
Comment (wasade) [inline skbio/stats/distance/_base.py:875]: Would it be possible to add in unit tests for the various paths through this function?

**Review by wasade — CHANGES_REQUESTED** [review]

File: skbio/stats/distance/_base.py (hunk 1 above)
Comment (sfiligoi) [reply skbio/stats/distance/_base.py:875]: OK.
Note, however, that the code is already indirectly tested as part of the _validate unit test.

Comment (sfiligoi) [reply skbio/stats/distance/_base.py:875]: Added.

File: skbio/stats/distance/tests/test_base.py [hunk 2]
```diff
@@ -979,6 +1010,55 @@ def test_to_series_default_ids(self):
         exp = pd.Series([0.123], index=pd.Index([('0', '1')]))
         assert_series_almost_equal(series, exp)
 
+    def test_validate_asym_shape(self):
+        # first check it actually likes good matrices
+        data_good = np.array([[0., 42.], [42., 0.]])
+        data_sym, data_hollow = is_symmetric_and_hollow(data_good)
+        self.assertEqual(data_sym, True)
+        del data_sym
```
Comment (wasade) [inline skbio/stats/distance/tests/test_base.py:1018]: why all the `del`s?

File: CHANGELOG.md [hunk 3]
```diff
@@ -13,6 +13,8 @@
 ### Performance enhancements
 
 * Avoid an implicit data copy on construction of `DissimilarityMatrix` objects. 
+* Avoid validation on copy of `DissimilarityMatrix` and `DistanceMatrix` objects.
```
Comment (wasade) [inline CHANGELOG.md:16]: Would it be possible to add a reference to this PR?

```suggestion
* Avoid validation on copy of `DissimilarityMatrix` and `DistanceMatrix` objects, see [PR #1747](https://github.com/biocore/scikit-bio/pull/1747)
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -13,6 +13,8 @@
...
+* Use an optimized version of symmetry check in DistanceMatrix.
```
Comment (wasade) [inline CHANGELOG.md:17]: Same here

```suggestion
* Use an optimized version of symmetry check in DistanceMatrix, see [PR #1747](https://github.com/biocore/scikit-bio/pull/1747)
```

**Review by wasade — APPROVED** [review]
Two minor suggestions to commit if you agree and a general question

File: skbio/stats/distance/tests/test_base.py (hunk 2 above)
Comment (sfiligoi) [reply skbio/stats/distance/tests/test_base.py:1018]: Mostly logical consistency...
Making sure I do not use an old value sometime later in the code.

Comment (wasade) [reply skbio/stats/distance/tests/test_base.py:1018]: Ah, thanks :)

**Comment by wasade** [comment]
@sfiligoi there are some conflicts here which appear complex. This may also conflict with #1749. Preference for order of merge?

**Comment by sfiligoi** [comment]
@wasade Merged, solving the conflicts.

**Pull request merged**